File: sdk-api/src/test/java/dev/restate/sdk/DeferredTest.java
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk;

import static dev.restate.sdk.JavaBlockingTests.*;

import dev.restate.sdk.common.Serde;
import dev.restate.sdk.common.StateKey;
import dev.restate.sdk.core.DeferredTestSuite;
import dev.restate.sdk.core.TestDefinitions.TestInvocationBuilder;
import dev.restate.sdk.serde.jackson.JsonSerdes;
import java.time.Duration;
import java.util.concurrent.TimeoutException;

public class DeferredTest extends DeferredTestSuite {

  protected TestInvocationBuilder reverseAwaitOrder() {
    return testDefinitionForVirtualObject(
        "ReverseAwaitOrder",
        Serde.VOID,
        JsonSerdes.STRING,
        (context, unused) -> {
          Awaitable<String> a1 = callGreeterGreetService(context, "Francesco");
          Awaitable<String> a2 = callGreeterGreetService(context, "Till");

          String a2Res = a2.await();
          context.set(StateKey.of("A2", JsonSerdes.STRING), a2Res);

          String a1Res = a1.await();

          return a1Res + "-" + a2Res;
        });
  }

  protected TestInvocationBuilder awaitTwiceTheSameAwaitable() {
    return testDefinitionForService(
        "AwaitTwiceTheSameAwaitable",
        Serde.VOID,
        JsonSerdes.STRING,
        (context, unused) -> {
          Awaitable<String> a = callGreeterGreetService(context, "Francesco");

          return a.await() + "-" + a.await();
        });
  }

  protected TestInvocationBuilder awaitAll() {
    return testDefinitionForService(
        "AwaitAll",
        Serde.VOID,
        JsonSerdes.STRING,
        (context, unused) -> {
          Awaitable<String> a1 = callGreeterGreetService(context, "Francesco");
          Awaitable<String> a2 = callGreeterGreetService(context, "Till");

          Awaitable.all(a1, a2).await();

          return a1.await() + "-" + a2.await();
        });
  }

  protected TestInvocationBuilder awaitAny() {
    return testDefinitionForService(
        "AwaitAny",
        Serde.VOID,
        JsonSerdes.STRING,
        (context, unused) -> {
          Awaitable<String> a1 = callGreeterGreetService(context, "Francesco");
          Awaitable<String> a2 = callGreeterGreetService(context, "Till");

          return (String) Awaitable.any(a1, a2).await();
        });
  }

  protected TestInvocationBuilder combineAnyWithAll() {
    return testDefinitionForService(
        "CombineAnyWithAll",
        Serde.VOID,
        JsonSerdes.STRING,
        (ctx, unused) -> {
          Awaitable<String> a1 = ctx.awakeable(JsonSerdes.STRING);
          Awaitable<String> a2 = ctx.awakeable(JsonSerdes.STRING);
          Awaitable<String> a3 = ctx.awakeable(JsonSerdes.STRING);
          Awaitable<String> a4 = ctx.awakeable(JsonSerdes.STRING);

          Awaitable<Object> a12 = Awaitable.any(a1, a2);
          Awaitable<Object> a23 = Awaitable.any(a2, a3);
          Awaitable<Object> a34 = Awaitable.any(a3, a4);
          Awaitable.all(a12, a23, a34).await();

          return a12.await() + (String) a23.await() + a34.await();
        });
  }

  protected TestInvocationBuilder awaitAnyIndex() {
    return testDefinitionForService(
        "AwaitAnyIndex",
        Serde.VOID,
        JsonSerdes.STRING,
        (ctx, unused) -> {
          Awaitable<String> a1 = ctx.awakeable(JsonSerdes.STRING);
          Awaitable<String> a2 = ctx.awakeable(JsonSerdes.STRING);
          Awaitable<String> a3 = ctx.awakeable(JsonSerdes.STRING);
          Awaitable<String> a4 = ctx.awakeable(JsonSerdes.STRING);

          return String.valueOf(Awaitable.any(a1, Awaitable.all(a2, a3), a4).awaitIndex());
        });
  }

  protected TestInvocationBuilder awaitOnAlreadyResolvedAwaitables() {
    return testDefinitionForService(
        "AwaitOnAlreadyResolvedAwaitables",
        Serde.VOID,
        JsonSerdes.STRING,
        (ctx, unused) -> {
          Awaitable<String> a1 = ctx.awakeable(JsonSerdes.STRING);
          Awaitable<String> a2 = ctx.awakeable(JsonSerdes.STRING);

          Awaitable<Void> a12 = Awaitable.all(a1, a2);
          Awaitable<Void> a12and1 = Awaitable.all(a12, a1);
          Awaitable<Void> a121and12 = Awaitable.all(a12and1, a12);

          a12and1.await();
          a121and12.await();

          return a1.await() + a2.await();
        });
  }

  protected TestInvocationBuilder awaitWithTimeout() {
    return testDefinitionForService(
        "AwaitOnAlreadyResolvedAwaitables",
        Serde.VOID,
        JsonSerdes.STRING,
        (ctx, unused) -> {
          Awaitable<String> call = callGreeterGreetService(ctx, "Francesco");

          String result;
          try {
            result = call.await(Duration.ofDays(1));
          } catch (TimeoutException e) {
            result = "timeout";
          }

          return result;
        });
  }
}

File: sdk-api-kotlin/src/test/kotlin/dev/restate/sdk/kotlin/DeferredTest.kt
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk.kotlin

import dev.restate.sdk.core.DeferredTestSuite
import dev.restate.sdk.core.TestDefinitions.*
import dev.restate.sdk.core.TestSerdes
import dev.restate.sdk.kotlin.KotlinCoroutinesTests.Companion.callGreeterGreetService
import dev.restate.sdk.kotlin.KotlinCoroutinesTests.Companion.testDefinitionForVirtualObject
import dev.restate.sdk.types.StateKey
import java.util.stream.Stream

class DeferredTest : DeferredTestSuite() {
  override fun reverseAwaitOrder(): TestInvocationBuilder =
      testDefinitionForVirtualObject("ReverseAwaitOrder") { ctx, _: Unit ->
        val a1: Awaitable<String> = callGreeterGreetService(ctx, "Francesco")
        val a2: Awaitable<String> = callGreeterGreetService(ctx, "Till")

        val a2Res: String = a2.await()
        ctx.set(StateKey.of("A2", TestSerdes.STRING), a2Res)

        val a1Res: String = a1.await()
        "$a1Res-$a2Res"
      }

  override fun awaitTwiceTheSameAwaitable(): TestInvocationBuilder =
      testDefinitionForVirtualObject("AwaitTwiceTheSameAwaitable") { ctx, _: Unit ->
        val a = callGreeterGreetService(ctx, "Francesco")
        "${a.await()}-${a.await()}"
      }

  override fun awaitAll(): TestInvocationBuilder =
      testDefinitionForVirtualObject("AwaitAll") { ctx, _: Unit ->
        val a1 = callGreeterGreetService(ctx, "Francesco")
        val a2 = callGreeterGreetService(ctx, "Till")

        listOf(a1, a2).awaitAll().joinToString(separator = "-")
      }

  override fun awaitAny(): TestInvocationBuilder =
      testDefinitionForVirtualObject("AwaitAny") { ctx, _: Unit ->
        val a1 = callGreeterGreetService(ctx, "Francesco")
        val a2 = callGreeterGreetService(ctx, "Till")
        Awaitable.any(a1, a2).await() as String
      }

  private fun awaitSelect(): TestInvocationBuilder =
      testDefinitionForVirtualObject("AwaitSelect") { ctx, _: Unit ->
        val a1 = callGreeterGreetService(ctx, "Francesco")
        val a2 = callGreeterGreetService(ctx, "Till")
        select {
          a1.onAwait { it }
          a2.onAwait { it }
        }
      }

  override fun combineAnyWithAll(): TestInvocationBuilder =
      testDefinitionForVirtualObject("CombineAnyWithAll") { ctx, _: Unit ->
        val a1 = ctx.awakeable(TestSerdes.STRING)
        val a2 = ctx.awakeable(TestSerdes.STRING)
        val a3 = ctx.awakeable(TestSerdes.STRING)
        val a4 = ctx.awakeable(TestSerdes.STRING)

        val a12 = Awaitable.any(a1, a2)
        val a23 = Awaitable.any(a2, a3)
        val a34 = Awaitable.any(a3, a4)
        Awaitable.all(a12, a23, a34).await()

        a12.await().toString() + a23.await() as String + a34.await()
      }

  override fun awaitAnyIndex(): TestInvocationBuilder =
      testDefinitionForVirtualObject("AwaitAnyIndex") { ctx, _: Unit ->
        val a1 = ctx.awakeable(TestSerdes.STRING)
        val a2 = ctx.awakeable(TestSerdes.STRING)
        val a3 = ctx.awakeable(TestSerdes.STRING)
        val a4 = ctx.awakeable(TestSerdes.STRING)

        Awaitable.any(a1, Awaitable.all(a2, a3), a4).awaitIndex().toString()
      }

  override fun awaitOnAlreadyResolvedAwaitables(): TestInvocationBuilder =
      testDefinitionForVirtualObject("AwaitOnAlreadyResolvedAwaitables") { ctx, _: Unit ->
        val a1 = ctx.awakeable(TestSerdes.STRING)
        val a2 = ctx.awakeable(TestSerdes.STRING)
        val a12 = Awaitable.all(a1, a2)
        val a12and1 = Awaitable.all(a12, a1)
        val a121and12 = Awaitable.all(a12and1, a12)
        a12and1.await()
        a121and12.await()

        a1.await() + a2.await()
      }

  override fun awaitWithTimeout(): TestInvocationBuilder {
    return unsupported("This is a feature not available in sdk-api-kotlin")
  }

  override fun definitions(): Stream<TestDefinition> =
      Stream.concat(super.definitions(), super.anyTestDefinitions { awaitSelect() })
}
